sway-core/src/asm_generation/fuel/data_section.rs

```diff
@@ -95,45 +74,24 @@ impl Entry {
         name: Option<String>,
         padding: Option<Padding>,
     ) -> Entry {
-        // We have to do some painful special handling here for enums, which are tagged unions.
-        // This really should be handled by the IR more explicitly and is something that will
-        // hopefully be addressed by https://github.com/FuelLabs/sway/issues/2819#issuecomment-1256930392
-
-        // Is this constant a tagged union?
-        if constant.ty.is_struct(context) {
-            let field_tys = constant.ty.get_field_types(context);
-            if field_tys.len() == 2
-                && field_tys[0].is_uint(context)
-                && field_tys[1].is_union(context)
-            {
-                // OK, this looks very much like a tagged union enum, which is the only place
-                // we use unions (otherwise we should be generalising this a bit more).
-                if let ConstantValue::Struct(els) = &constant.value {
-                    if els.len() == 2 {
-                        let tag_entry = Entry::from_constant(context, &els[0], None, None);
-
-                        // Here's the special case.  We need to get the size of the union and
-                        // attach it to this constant entry which will be one of the variants.
-                        let val_entry = {
-                            let target_size = size_bytes_round_up_to_word_alignment!(
-                                ir_type_size_in_bytes(context, &field_tys[1]) as usize
-                            );
-                            Entry::from_constant(
-                                context,
-                                &els[1],
-                                None,
-                                Some(Padding::Left { target_size }),
-                            )
-                        };
-
-                        // Return here from our special case.
-                        return Entry::new_collection(vec![tag_entry, val_entry], name, padding);
-                    }
-                }
-            }
-        };
+        // We need a special handling in case of enums.
+        if constant.ty.is_enum(context) {
+            let tag_and_value_with_paddings = constant
+                .elements_of_aggregate_with_padding(context)
```

Comment:
This is MUCH better...
but why not naming this like `data_section_layout_as_enum` or something like that.

`elements_of_aggregate_with_padding` is little opaque to me.